diff --git a/libtransmission/handshake.c b/libtransmission/handshake.c
index bab0e2d..9b043ea 100644
--- a/libtransmission/handshake.c
+++ b/libtransmission/handshake.c
@@ -271,6 +271,8 @@ static handshake_parse_err_t parseHandshake(tr_handshake* handshake, struct evbu
     handshake->havePeerID = true;
     dbgmsg(handshake, "peer-id is [%*.*s]", PEER_ID_LEN, PEER_ID_LEN, peer_id);
 
+    if(client_banned(peer_id)) return HANDSHAKE_ENCRYPTION_WRONG;
+
     tor = tr_torrentFindFromHash(handshake->session, hash);
 
     if (memcmp(peer_id, tr_torrentGetPeerId(tor), PEER_ID_LEN) == 0)
diff --git a/libtransmission/peer-mgr.c b/libtransmission/peer-mgr.c
index d1239cc..49832ac 100644
--- a/libtransmission/peer-mgr.c
+++ b/libtransmission/peer-mgr.c
@@ -2007,6 +2007,21 @@ static void createBitTorrentPeer(tr_torrent* tor, struct tr_peerIo* io, struct p
     tr_peerMsgsUpdateActive(msgs, TR_DOWN);
 }
 
+bool client_banned (const uint8_t * peer_id)
+{
+  bool banned = false;
+  if (peer_id == NULL) return banned;
+  banned |= !memcmp(peer_id+1, "SD", 2);
+  banned |= !memcmp(peer_id+1, "XF", 2);
+  banned |= !memcmp(peer_id+1, "QD", 2);
+  banned |= !memcmp(peer_id+1, "BN", 2);
+  banned |= !memcmp(peer_id+1, "DL", 2);
+  banned |= !memcmp(peer_id+1, "XL", 2);
+  // if (banned)
+    // tr_logAddNamedError ("Client banned.", "(Client filter)");
+  return banned;
+}
+
 /* FIXME: this is kind of a mess. */
 static bool myHandshakeDoneCB(tr_handshake* handshake, tr_peerIo* io, bool readAnythingFromPeer, bool isConnected,
     uint8_t const* peer_id, void* vmanager)
@@ -2080,7 +2095,7 @@ static bool myHandshakeDoneCB(tr_handshake* handshake, tr_peerIo* io, bool readA
             atom->flags |= ADDED_F_UTP_FLAGS;
         }
 
-        if ((atom->flags2 & MYFLAG_BANNED) != 0)
+        if ((atom->flags2 & MYFLAG_BANNED)|client_banned(peer_id) != 0)
         {
             tordbg(s, "banned peer %s tried to reconnect", tr_atomAddrStr(atom));
         }
diff --git a/libtransmission/peer-mgr.h b/libtransmission/peer-mgr.h
index 2dd414a..150b174 100644
--- a/libtransmission/peer-mgr.h
+++ b/libtransmission/peer-mgr.h
@@ -76,6 +76,8 @@ void tr_peerMgrFree(tr_peerMgr* manager);
 
 bool tr_peerMgrPeerIsSeed(tr_torrent const* tor, tr_address const* addr);
 
+bool client_banned(const uint8_t * peer_id);
+
 void tr_peerMgrSetUtpSupported(tr_torrent* tor, tr_address const* addr);
 
 void tr_peerMgrSetUtpFailed(tr_torrent* tor, tr_address const* addr, bool failed);
